### PR TITLE
Componentizes cult halos + fixes some cult halo issues

### LINKED ISSUE #44146: cult nonhuman mobs dont get the cult halo overlay
this means you can have ape cultists in stealth mode which is probably not intentional

### LINKED ISSUE #57204: Cult changelings lose their halo and glowy eyes when they transform
## Round ID:
#157131

## Reproduction:

When a cultist changeling transforms, they lose their halo and eyes. When they transform from a cultist dna to a noncultist dna back to a cultist dna, they don't seem to regain their eyes or halos.

### LINKED ISSUE #58585: Cultist Halo vanishes
## Reproduction:
I was transformed into a monkey as cultist and after my return to humanity i lacked the halo and my glowing eyes.

### PR DESCRIPTION
**Pull request by SuperNovaa41**
## About The Pull Request
Fixes https://github.com/tgstation/tgstation/issues/58585
Fixes https://github.com/tgstation/tgstation/issues/57204
Fixes https://github.com/tgstation/tgstation/issues/44146
Turns the status of a cultist's visibility into a component.
The adding and removing of red eyes and halos are handled by the component.

## Why It's Good For The Game
Refactors cult code a bit, makes it far easier to manage the overlays and traits given by the size of the cult.
Fixes some exploits and issues.

## Changelog

:cl:
fix: You can no longer delete your cult eyes or halo by transforming into a monkey, and then into a human.
fix: You can no longer delete your cult eyes or halo by transforming as a changeling.
fix: Simplemobs now receive glowing eyes and a cult halo.
refactor: Turns a cultist's visibility status into a component, and improves cult code a bit.
/:cl:

### CODE REVIEW

File: code/datums/components/cult_status.dm [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+/datum/component/cult_status/proc/set_eyes()
+	if (ishuman(parent))
+		var/mob/living/carbon/human/human_parent = parent
+		human_parent.eye_color = BLOODCULT_EYE
+		human_parent.dna.update_ui_block(DNA_EYE_COLOR_BLOCK)
+		ADD_TRAIT(human_parent, TRAIT_UNNATURAL_RED_GLOWY_EYES, CULT_TRAIT)
+		human_parent.update_body()
+	else
+		var/mob/living/simple_animal/simple_parent = parent
```
Comment (Fikou) [inline code/datums/components/cult_status.dm:64]: you dont need to set type here at all, and this isnt guarranteed to be a simplemob

**Review by MrMelbert — CHANGES_REQUESTED** [review]
It's a good first attempt at a component, but there's a few ickys around.

Signals should generally only be sent from one place ever - they're meant to report an event happened, not as a pseudo proc-call. A few of your signals definitely in that trap. 

The component itself needs some auto-doc comments as well.

File: code/modules/antagonists/cult/cult.dm [hunk 2]
```diff
@@ -138,9 +140,9 @@
 		magic.Grant(current)
 	current.throw_alert("bloodsense", /atom/movable/screen/alert/bloodsense)
 	if(cult_team.cult_risen)
-		cult_team.rise(current)
+		SEND_SIGNAL(current, COMSIG_CULT_VIS, STAGE_CULT_RED_EYES, override_timer = TRUE)
 		if(cult_team.cult_ascendent)
-			cult_team.ascend(current)
+			SEND_SIGNAL(current, COMSIG_CULT_VIS, STAGE_CULT_HALOS, override_timer = TRUE)
```
[the remark this answers is not in the document]
Comment (SuperNovaa41) [reply code/modules/antagonists/cult/cult.dm:145]: this is the only place this logic is handled this way. the other places that `COMSIG_CULT_VIS` is called would not make sense to do it like that

File: code/datums/components/cult_status.dm [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * Check if the cultist needs to have their
+ */
+/datum/component/cult_status/Initialize()
+	RegisterSignal(parent, COMSIG_CULT_VIS, .proc/raise_level)
+	RegisterSignal(parent, COMSIG_ANTAGONIST_REMOVED, .proc/UnregisterFromParent)
+	RegisterSignal(parent, COMSIG_MOB_TRANSFORMING, .proc/handle_transform)
+
+	var/datum/team/cult/parent_team
+	for (var/datum/antagonist/cult/cult_datum in parent.mind.antag_datums)
```
Comment (JohnFulpWillard) [inline code/datums/components/cult_status.dm:25]:
```suggestion
	var/mob/living/carbon/carbon_parent = parent
	for (var/datum/antagonist/cult/cult_datum in carbon_parent.mind.antag_datums)
```

File: code/modules/antagonists/cult/cult.dm [hunk 4]
```diff
@@ -158,11 +155,8 @@
 	current.clear_alert("bloodsense")
 	if(ishuman(current))
 		var/mob/living/carbon/human/H = current
-		H.eye_color = initial(H.eye_color)
-		H.dna.update_ui_block(DNA_EYE_COLOR_BLOCK)
+		SEND_SIGNAL(H, COMSIG_CULT_DECONVERT)
```
Comment (JohnFulpWillard) [inline code/modules/antagonists/cult/cult.dm:158]: maybe this can just be an 'update_level' signal instead of only having 'raise_level'

**Review by JohnFulpWillard — COMMENTED** [review]
i wish we fixed glowing eyes not showing up on examine when you wear one-sided huds but i guess this is a better fix atm

File: code/datums/components/cult_status.dm [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	///Have the eyes started glowing?
+	var/glowing_eyes = FALSE
+	///Has the halo been added?
+	var/has_halo = FALSE
+
+/**
+ * Check if the cultist needs to have their
+ */
+/datum/component/cult_status/Initialize()
+	RegisterSignal(parent, COMSIG_CULT_VIS, .proc/raise_level)
```
Comment (TemporalOroboros) [inline code/datums/components/cult_status.dm:20]: Procnals are generally discouraged. Could you make this a global signal when the status of the cult changes and have the halos decide what stage to take?

**Comment by SuperNovaa41** [comment]
Yeah, just going to close this and redo it as an element.

**Comment by JohnFulpWillard** [comment]
:(

**Pull request closed without merging**